File: js/calc.js
/**
 * calc.js
 * Core financial calculation engine for the Mortgage Switch Calculator.
 *
 * All rate inputs are already decimals (e.g. 0.0875 for 8.75%).
 * "Effective rate" for index-linked loans = nominal rate + inflation.
 */

/* ─────────────────────────────────────────────────────────────────────────────
   HELPERS
───────────────────────────────────────────────────────────────────────────── */

/**
 * Compute the effective annual rate.
 * For index-linked loans the inflation premium is added to the nominal rate.
 *
 * @param {number} nominalRate  - Annual nominal rate (decimal)
 * @param {string} type         - "fixed" | "variable" | "index"
 * @param {number} inflation    - Annual inflation rate (decimal), used for "index"
 * @returns {number} Effective annual rate (decimal)
 */
export function effectiveRate(nominalRate, type, inflation) {
  return type === 'index' ? nominalRate + inflation : nominalRate;
}

/**
 * Monthly annuity payment given principal, annual rate, and term in years.
 *
 * @param {number} principal
 * @param {number} annualRate   - decimal (e.g. 0.0875)
 * @param {number} totalYears
 * @returns {number} Monthly payment amount
 */
export function monthlyPayment(principal, annualRate, totalYears) {
  if (annualRate <= 0) return principal / (totalYears * 12);
  const r = annualRate / 12;
  const n = totalYears * 12;
  return principal * r * Math.pow(1 + r, n) / (Math.pow(1 + r, n) - 1);
}

/**
 * Estimate total interest paid over the full life of a loan.
 *
 * @param {number} principal
 * @param {number} effRate    - Effective annual rate (decimal)
 * @param {number} totalYears
 * @returns {number} Total interest
 */
export function totalInterestEstimate(principal, effRate, totalYears) {
  const mp = monthlyPayment(principal, effRate, totalYears);
  return mp * totalYears * 12 - principal;
}

/* ─────────────────────────────────────────────────────────────────────────────
   SINGLE SWITCH
───────────────────────────────────────────────────────────────────────────── */

/**
 * Calculate the net position for every month from 0 → totalMonths if the
 * borrower switches to Option A at that month.
 *
 * @param {object} p - inputs object (from inputs.js getInputs())
 * @returns {object} {
 *   rows[]          - {m, switchCost, savings, net, yearsRemaining}
 *   bestMonth       - month at which net is maximised
 *   bestNet         - maximum net value
 *   firstPositive   - first month where net > 0  (-1 if never)
 *   totalMonths     - p.yearsLeft × 12
 *   monthlySaving   - monthly saving after switch
 *   effCurrent      - effective current rate
 *   effNew          - effective Option A rate
 * }
 */
export function calcSingleSwitch(p) {
  const totalMonths   = Math.round(p.yearsLeft * 12);
  const effCurrent    = effectiveRate(p.currentRate,  p.currentType,  p.inflationCurrent);
  const effNew        = effectiveRate(p.newRate,       p.newRateType,  p.inflationA);
  const monthlySaving = p.principal * (effCurrent - effNew) / 12;

  const rows = [];
  let bestMonth = 0, bestNet = -Infinity, firstPositive = -1;

  for (let m = 0; m <= totalMonths; m++) {
    const yearsRemaining = p.yearsLeft - m / 12;
    const penalty        = p.principal * p.penaltyRate * Math.max(yearsRemaining, 0);
    const switchCost     = penalty + p.setupFee;
    const savings        = monthlySaving * (totalMonths - m);
    const net            = savings - switchCost;

    if (net > bestNet)              { bestNet = net; bestMonth = m; }
    if (net > 0 && firstPositive === -1) { firstPositive = m; }

    rows.push({ m, switchCost, savings, net, yearsRemaining });
  }

  return { rows, bestMonth, bestNet, firstPositive, totalMonths, monthlySaving, effCurrent, effNew };
}

/* ─────────────────────────────────────────────────────────────────────────────
   DOUBLE SWITCH  (A now → B at month M2)
───────────────────────────────────────────────────────────────────────────── */

/**
 * Calculate the cumulative net position for the double-switch strategy:
 *   1. Switch to Option A immediately (incurring cost1)
 *   2. At month M2, break Option A and switch to Option B (incurring cost2)
 *
 * @param {object} p - inputs object
 * @returns {object|null} null if M2 >= totalMonths
 */
export function calcDoubleSwitch(p) {
  const totalMonths = Math.round(p.yearsLeft * 12);
  const M2          = p.switchMonth2;

  if (M2 >= totalMonths) return null;

  const effCurrent = effectiveRate(p.currentRate, p.currentType,  p.inflationCurrent);
  const effNew     = effectiveRate(p.newRate,      p.newRateType,  p.inflationA);
  const effNew2    = effectiveRate(p.rate2,        p.rateType2,    p.inflationB);

  const saving1 = p.principal * (effCurrent - effNew)  / 12;
  const saving2 = p.principal * (effCurrent - effNew2) / 12;

  // Cost of first switch (today)
  const cost1 = p.principal * p.penaltyRate * p.yearsLeft + p.setupFee;

  // Penalty for breaking Option A at month M2
  const fixedTermA       = p.newRateType === 'fixed' ? p.newFixedTerm : p.yearsLeft;
  const yearsLeftOnA     = fixedTermA - M2 / 12;
  const penalty2         = yearsLeftOnA > 0
    ? p.principal * p.penaltyRateNew * yearsLeftOnA
    : 0;
  const cost2            = penalty2 + p.setupFee2;

  const phase1Savings  = saving1 * M2;
  const phase2Savings  = saving2 * (totalMonths - M2);
  const totalNet       = phase1Savings + phase2Savings - cost1 - cost2;

  // Build month-by-month cumulative timeline
  const rows = [];
  let cum = -cost1;

  for (let m = 0; m <= totalMonths; m++) {
    if (m === M2) cum -= cost2;
    rows.push({ m, net: cum });
    const rate = m < M2 ? saving1 : saving2;
    if (m < totalMonths) cum += rate;
  }

  return {
    M2, cost1, cost2, penalty2, yearsLeftOnA,
    phase1Savings, phase2Savings, totalNet,
    saving1perMonth: saving1,
    saving2perMonth: saving2,
    rows, totalMonths,
    effCurrent, effNew, effNew2
  };
}

/* ─────────────────────────────────────────────────────────────────────────────
   SCENARIO BUILDER
───────────────────────────────────────────────────────────────────────────── */

/**
 * Build a list of all comparable strategies with their net outcomes.
 *
 * @param {object}      p           - inputs
 * @param {object}      single      - result of calcSingleSwitch(p)
 * @param {object|null} dbl         - result of calcDoubleSwitch(p), or null
 * @param {number}      totalMonths
 * @returns {Array<{label, net, color, cost}>}
 */
export function buildScenarios(p, single, dbl, totalMonths) {
  const scenarios = [];

  // Baseline
  scenarios.push({ label: 'Stay on current rate', net: 0, color: 'var(--muted)', cost: 0 });

  // Switch A only
  if (single.bestNet !== -Infinity) {
    scenarios.push({
      label: 'Switch to A now, hold',
      net:   single.bestNet,
      color: 'var(--accent3)',
      cost:  single.rows[0].switchCost,
    });
  }

  if (dbl) {
    // Switch A now → B later
    scenarios.push({
      label: `Switch A now → B at month ${dbl.M2}`,
      net:   dbl.totalNet,
      color: 'var(--purple)',
      cost:  dbl.cost1 + dbl.cost2,
    });

    // Wait and switch to B only (skip A)
    const M2            = dbl.M2;
    const effCurrent    = effectiveRate(p.currentRate, p.currentType, p.inflationCurrent);
    const effB          = effectiveRate(p.rate2, p.rateType2, p.inflationB);
    const savingBOnly   = p.principal * (effCurrent - effB) / 12;
    const yearsLeftAtM2 = Math.max(p.yearsLeft - M2 / 12, 0);
    const penaltyBOnly  = yearsLeftAtM2 * p.principal * p.penaltyRate;
    const costBOnly     = penaltyBOnly + p.setupFee2;
    const netBOnly      = savingBOnly * (totalMonths - M2) - costBOnly;

    scenarios.push({
      label: `Wait → switch to B only at month ${M2}`,
      net:   netBOnly,
      color: 'var(--accent2)',
      cost:  costBOnly,
    });
  }

  return scenarios;
}

File: reporter/calc.py
"""
calc.py
Python port of the core financial logic from js/calc.js.
All rate inputs are decimals (e.g. 0.0875 for 8.75%).
"""

import math


def effective_rate(nominal_rate: float, loan_type: str, inflation: float) -> float:
    """Effective annual rate. For index-linked loans, inflation is added."""
    return nominal_rate + inflation if loan_type == "index" else nominal_rate


def monthly_payment(principal: float, annual_rate: float, total_years: float) -> float:
    """Monthly annuity payment."""
    if annual_rate <= 0:
        return principal / (total_years * 12)
    r = annual_rate / 12
    n = total_years * 12
    return principal * r * math.pow(1 + r, n) / (math.pow(1 + r, n) - 1)


def total_interest_estimate(principal: float, eff_rate: float, total_years: float) -> float:
    mp = monthly_payment(principal, eff_rate, total_years)
    return mp * total_years * 12 - principal


def remaining_balance(
    principal: float, annual_rate: float, total_years: float, months_elapsed: int
) -> float:
    """Outstanding principal after months_elapsed payments on a standard annuity."""
    if months_elapsed <= 0:
        return principal
    if annual_rate <= 0:
        n = round(total_years * 12)
        return max(0.0, principal * (1 - months_elapsed / n))
    r = annual_rate / 12
    n = round(total_years * 12)
    mp = principal * r * math.pow(1 + r, n) / (math.pow(1 + r, n) - 1)
    return max(
        0.0,
        principal * math.pow(1 + r, months_elapsed)
        - mp * (math.pow(1 + r, months_elapsed) - 1) / r,
    )


def remaining_balance_indexed(
    principal: float,
    real_rate: float,
    inflation: float,
    total_years: float,
    months_elapsed: int,
) -> float:
    """Outstanding nominal principal for an index-linked loan."""
    if months_elapsed <= 0:
        return principal
    real_remaining = remaining_balance(principal, real_rate, total_years, months_elapsed)
    return real_remaining * math.pow(1 + inflation, months_elapsed / 12)


def amortization_schedule(
    principal: float, annual_rate: float, total_years: int, loan_type: str, inflation: float
) -> list[dict]:
    """
    Full month-by-month amortization schedule.
    Returns a list of dicts with keys:
        month, payment, interest, principal_paid, balance
    For index-linked loans, balance grows in nominal terms in early years.
    """
    eff = effective_rate(annual_rate, loan_type, inflation)
    payment = monthly_payment(principal, eff, total_years)
    schedule = []
    balance = principal
    monthly_inf = (1 + inflation) ** (1 / 12) - 1 if loan_type == "index" else 0.0
    monthly_rate = annual_rate / 12

    for m in range(1, total_years * 12 + 1):
        if loan_type == "index":
            # Index the balance first, then apply interest and payment
            balance *= 1 + monthly_inf
            interest = balance * monthly_rate
        else:
            interest = balance * monthly_rate

        principal_paid = payment - interest
        balance = max(0.0, balance - principal_paid)
        schedule.append({
            "month": m,
            "payment": round(payment),
            "interest": round(interest),
            "principal_paid": round(principal_paid),
            "balance": round(balance),
        })

    return schedule


def break_even_analysis(config: dict, offer: dict) -> dict:
    """
    Compute the net savings and break-even month if switching from the
    current loan (config) to offer today.

    Returns:
        {
            "monthly_saving":   float,   # positive = cheaper on new loan
            "break_even_month": int|None,
            "net_at_term_end":  float,
            "switch_cost":      float,
        }
    """
    principal = config["principal"]
    years_left = config["years_left"]
    total_term = config["total_term"]
    inflation_current = config.get("inflation", 0.0)
    penalty_rate = config.get("penalty_rate", 0.0)
    setup_fee = config.get("setup_fee", 0.0)

    eff_current = effective_rate(config["annual_rate"], config["loan_type"], inflation_current)
    eff_new = effective_rate(offer["annual_rate"], offer["loan_type"], inflation_current)

    mp_current = monthly_payment(principal, eff_current, total_term)
    mp_new = monthly_payment(principal, eff_new, total_term)
    monthly_saving = mp_current - mp_new

    switch_cost = principal * penalty_rate * years_left + setup_fee
    total_months = round(years_left * 12)

    best_net = -math.inf
    break_even_month = None

    for m in range(total_months + 1):
        if config["loan_type"] == "index":
            p_m = remaining_balance_indexed(
                principal, config["annual_rate"], inflation_current, total_term, m
            )
        else:
            p_m = remaining_balance(principal, eff_current, total_term, m)

        penalty = p_m * penalty_rate * max(years_left - m / 12, 0)
        cost = penalty + setup_fee
        rem_years = max(total_term - m / 12, 1 / 12)
        m_saving = monthly_payment(p_m, eff_current, rem_years) - monthly_payment(p_m, eff_new, rem_years)
        savings = m_saving * (total_months - m)
        net = savings - cost

        if net > best_net:
            best_net = net

        if net > 0 and break_even_month is None:
            break_even_month = m

    return {
        "monthly_saving": round(monthly_saving),
        "break_even_month": break_even_month,
        "net_at_term_end": round(best_net),
        "switch_cost": round(switch_cost),
    }
